Connection strategy changeover - pre 0.5 blocker clearing

NoraCodes opened the pull request:
Yay, here's the big refactor and test block that gives us working database connections! In essence it moves the connection responsibility from the point of use to the point of call; e.g. data base connections are no longer generated in models but rather are generated in routes, in tests, and in scripts.

This gives a lot more control over which DB connection is used.

Once this is merged 0.5 is not far behind. This includes all code for, and thus obviates and closes #399, closes #347, closes #363, closes #350, and closes #335

Also closes #405

simonbtomlinson committed:
Create materialized views that aggregate readings

simonbtomlinson committed:
Add a minute precision view

simonbtomlinson committed:
Script to refresh reading views

simonbtomlinson committed:
Better algorithm for compressed readings

simonbtomlinson committed:
Partial materializion strategy for compression

Adds views for compressed data at the minute / hour / day level, and
designates the day one as a materialized view. All these views should
only be accessed through a function (currently named
compressed_readings_2) that will only query the fine precision ones for
short intervals, negating their slowness. The daily view is materialized
to improve its performance.

Refreshing of the materialized view is not yet implemented.

simonbtomlinson committed:
Fix bad extract() syntax

simonbtomlinson committed:
Add creation of btree gist extension

simonbtomlinson committed:
Add permissions for the btree_gist extension

The OED user didn't have superuser permissions, so they couldn't create
the btree_gist extension.

This also removes an sql function that refreshed a materialized view,
because functions aren't allowed to do that. The refresh code is now
inlined in Reading.js.

simonbtomlinson committed:
Add tests for compressed readings

simonbtomlinson committed:
Remove create_compressed_readings_faster.sql

create_cmporessed_readings_faster.sql contained an old version of the
code for faster compressed readings. The logic there is now in
create_compressed_reading_views.sql and
generateCompressedReadingViews.js.

simonbtomlinson committed:
New compressed readings for groups

simonbtomlinson committed:
Test for compressed group readings

simonbtomlinson committed:
Compressed barchart readings

simonbtomlinson committed:
Fix logging in refreshReadingViews script

simonbtomlinson committed:
Fix a bug in compressed barchart readings

Compressed barchart readings was returning results in kilojoules instead
of kwh because of a bad unit conversion.

Also wrote a unit test to capture the bug.

simonbtomlinson committed:
Test for multi-day meter barchart compression

simonbtomlinson committed:
Test for barchart with multiple meters

simonbtomlinson committed:
Begin adding routes for new compressed readings

simonbtomlinson committed:
Handle infinite timestamps for new compressed readings

simonbtomlinson committed:
Compressed route for group line readings

simonbtomlinson committed:
Compressed route for meter bar readings

simonbtomlinson committed:
Compressed route for group bar readings

simonbtomlinson committed:
Client-side api for compressed readings

simonbtomlinson committed:
Client-side implementation excluding export functionality

simonbtomlinson committed:
Merge branch 'development' into fast-pt-compression
Export and compare are currently broken

simonbtomlinson committed:
Fix export

simonbtomlinson committed:
Merge branch 'development' into fast-pt-compression

simonbtomlinson committed:
Fix lint errors

simonbtomlinson committed:
Merge branch 'development' into fast-pt-compression

simonbtomlinson committed:
SQL for meter compare calculations

simonbtomlinson committed:
SQL for group compare

simonbtomlinson committed:
Api endpoint for compare

NoraCodes committed:
Add dual toggle for meters.

Meters can now be enabled or not (collecting data) and visible or not
(shown to users). There are also methods (and tests) to get only the
meters marked with the appropriate flags.

Meters are enabled and visible by default.

NoraCodes committed:
Use only enabled meters in updateMamacMeters.

Commit:
Fix a test error for creating btree_gist extension
Copied from baseline_refactoring branch fe5f115 and 115f150.

Commit:
Merge development into fast-pt-compression

Commit:
Add a missing comma in package.json

Commit:
Use getDB func for compressed reading tests

Commit:
Revert "Fix a test error for creating btree_gist extension"

This reverts commit 21f11836cb0c74ec7cf8b6c013e5ff83243d93ff.

Commit:
Revert "Revert "Fix a test error for creating btree_gist extension""

This reverts commit ab8e87b769c9b56a5d54cc26198aec565364c7b4.

Commit:
Add a missing semicolon to a sql file

NoraCodes committed:
Filter meters by authentication status

Now, non-authenticated users can't see non-displayable meters

NoraCodes committed:
UI for meters toggling; display only

Provides a tabular view of all meters in the system, with all associated
information (including toggles). Users who are not logged in see only
meters with the Displayable flag on, regardless of what view is in use,
and the server does not send IP or meter type in the absence of a token.

NoraCodes committed:
Fix TSC/ATLoader not giving errors in some cases

ATLoader needs to have a plugin enabled to give good error messages and
we hadn't done that, so here it is.

NoraCodes committed:
Add buttons for editing meter toggles. Not active.

NoraCodes committed:
Meter editing from the interface.

Adds new actions and improves error handling for authentication.

Commit:
Add missing await for a migration test

Commit:
Merge branch 'development' into fast-pt-compression

Commit:
Update packages to fix the vulnerability of cryptiles

Commit:
Add a bash script for refreshReadingViews

Commit:
Update nodemon to avoid npm install error caused by removed package, flapmap-stream

Commit:
Filter refreshReadingViews script out of checks for MPL header

Commit:
Correct a file name in checkHeader

Commit:
Add descriptions to refreshReadingViews script

NoraCodes committed:
Fix for groups button missing

NoraCodes committed:
Merge development to meter-update-display-optional

NoraCodes committed:
Fix meter edit button getting squished

NoraCodes committed:
Improve meters list for unauthenticated users.

Fix button size, remove button and adming columns for unauthed users.

NoraCodes committed:
Fix linting failures.

Commit:
Merge branch 'development' into fast-pt-compression

Commit:
Modify compare_readings SQL functions to calculate with time shift

Commit:
Fix compare_readings functions and lint errors

Commit:
Remove console log messages

Commit:
Add refreshReadingsView to package.json

Commit:
Change file name for compare_readings SQL function

Commit:
Add get Start/End Timestamp functions to TimeInterval

Commit:
Modify state for comparison graph

Commit:
Modify actions for comparison graph

Commit:
Modify Redux part of Comparison graph

Commit:
Modify React part of Comparison graph

Commit:
Add missing Mozilla License headers

Commit:
Get timeInterval and compareShift strings only when action type is valid

Commit:
Use curr_use and prev_use for both front-end and back-end of comparison

Commit:
Set migration for compare_readings/group_compare_readings sql functions

Commit:
remove console log

Commit:
Fix Reading model to use curr_use instead of current_use

Commit:
Fix CompareChartContainer to support undefind prevTotalUsage data

Commit:
Remove unused lines

Commit:
Remove prev total usage bar/reading

shuheif committed:
Modified chart labels

shuheif committed:
Fix a linting error

TristanVaca committed:
First steps toward getDB going away

NoraCodes committed:
Made migration invalid tests acutally test stuff

NoraCodes committed:
Fixed problem with reference export for test db

Look at groups tests for an archetype of how to use the new method.

NoraCodes committed:
Document exports from test/db/common.js

NoraCodes committed:
Change connection strategy for groups tests.

NoraCodes committed:
Fix readings and BarchartReadingsTests for conn

NoraCodes committed:
Fix linting failure

NoraCodes committed:
Update baseline to new conn technique

NoraCodes committed:
Update compressed readings with new conn strategy

NoraCodes committed:
Update cyclic group tests for new conn strategy

NoraCodes committed:
Update meters-update-display-optional<-development

Including reconciling package-lock.json, 0.3.0 to 0.4.0 migration, and
making the baseline tests use the right Meter constructor.

NoraCodes committed:
Merge branch 'development' of github.com:OpenEnergyDashboard/OED into meter-update-display-optional

NoraCodes committed:
Update loadFromCsvStream and tests for conn

NoraCodes committed:
Move raw SQL in readings tests to model

NoraCodes committed:
Update csv stream tests to new conn strategy

NoraCodes committed:
Update log email tests to new conn strategy

NoraCodes committed:
Update metasys tests to new conn strategy

and consolidate metasys tests into a single file, because, like, why.

NoraCodes committed:
Change return type of Reading.count

NoraCodes committed:
Fix wrong number of args in barchart test

NoraCodes committed:
Update meters tests to new conn strategy

NoraCodes committed:
Don't try to send log emails from tests

NoraCodes committed:
Update Moment patch tests to new conn strategy

NoraCodes committed:
Update reading tests to new conn strategy

NoraCodes committed:
Remove test modification metadata

NoraCodes committed:
Fix migrations tests to actually work

NoraCodes committed:
Fix linting errors

NoraCodes committed:
Initial work for API tests in test/web/

NoraCodes committed:
First working HTTP API test: version

This required creating a HTTP API tests folder, src/server/test/web, and
adding chai-http as a dependency. In addition I had to change linting
rules and disable logging when inappropriate.

NoraCodes committed:
Add test for meters API

This includes adding a test user to the database on each recreateDB.

NoraCodes committed:
Add ultra basic tests for readings API

Also, fix a bug in readings API where almost everything would be accepted through the validation step.

NoraCodes committed:
Add tests for login API

NoraCodes committed:
Fix lint errors in API tests

NoraCodes committed:
Add tests for groups retrieval API.

NoraCodes committed:
Fix bug where only one deep child was returned

NoraCodes committed:
Add auth tests for groups create/edit API

NoraCodes committed:
Fix bug where auth would pass on after failure

NoraCodes committed:
Always log to the file and log promise rejections

NoraCodes committed:
Add tests for groups create and modify API

These tests directly add a new meter or group to the database, and then
use the API to add it as a child to a group, and then again to verify
that it was actually added.

NoraCodes committed:
Fix bug where API didn't support group add/edit

The wrong type of connection object was being passed to the model
function for the transaction; now, it works, though this work should be
erased soon with the new conn model.

NoraCodes committed:
Merge branch 'development' into api-tests

NoraCodes committed:
Basic union of api test and conn changes

NoraCodes committed:
Update 'api-tests' to development

NoraCodes committed:
Merge working copies for api-tests

NoraCodes committed:
Finished conn strategy swapover.

Some migrations tests still fail.
In essence, there is now more fetching of database connections, at the
highest context possible.

NoraCodes committed:
Switch misc. scripts to new conn strategy

NoraCodes committed:
Set up debugger passthrough for Node from Docker

NoraCodes committed:
Fix linting errors.

huss commented on src/server/log.js:
```diff
@@ -2,9 +2,12 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+// TODO: This subsystem should make its own database connections.
+
```
Is this okay and comment correct?

huss commented on src/server/models/Group.js:
```diff
@@ -23,21 +22,21 @@ class Group {
 	 */
-	static createTables() {
-		return getDB().none(sqlFile('group/create_groups_tables.sql'));
+	static createTables(conn) {
```
Should there be a comment about conn is one used as in other places?

huss commented on src/server/models/Meter.js:
```diff
@@ -28,37 +27,37 @@ class Meter {
 	 */
-	static createTable() {
-		return getDB().none(sqlFile('meter/create_meters_table.sql'));
+	static createTable(conn) {
```
@param for conn?

huss commented on src/server/models/Meter.js:
```diff
@@ -28,37 +27,37 @@ class Meter {
 	 */
-	static createMeterTypesEnum() {
-		return getDB().none(sqlFile('meter/create_meter_types_enum.sql'));
+	static createMeterTypesEnum(conn) {
```
@param for conn?

huss commented on src/server/models/Preferences.js:
```diff
@@ -24,28 +23,33 @@ class Preferences {
 	}
 
-	static async update(newPreferences) {
+	static async update(conn, newPreferences) {
```
@param for conn?

huss commented on src/server/models/Reading.js:
```diff
@@ -26,85 +25,95 @@ class Reading {
 	 */
-	static createTable() {
-		return getDB().none(sqlFile('reading/create_readings_table.sql'));
+	static createTable(conn) {
```
@param for conn for following ones?

huss commented on src/server/models/User.js:
```diff
@@ -21,60 +20,66 @@ class User {
 	 */
-	static async updateUserPassword(email, passwordHash) {
-		return getDB().none(sqlFile('user/update_user_password.sql'), { email: email, password_hash: passwordHash });
+	static async updateUserPassword(conn, email, passwordHash) {
```
I noticed conn is the first param here but last in other cases.

huss commented on src/server/routes/users.js:
```diff
@@ -40,6 +42,7 @@ router.get('/:user_id', async (req, res) => {
 	if (!validate(req.params, validParams).valid) {
 		res.sendStatus(400);
 	} else {
+		const conn = getConnection();
```
I'm uncertain where this conn is used.

kuchtact replied:
It is used on line 47.

huss commented on src/server/test/db/compressedReadingsTests.js:
```diff
@@ -50,40 +44,40 @@ mocha.describe('Compressed Readings', () => {
 		const reading1 = new Reading(meter.id, 100, timestamp1, timestamp2);
 		const reading2 = new Reading(meter.id, 200, timestamp2, timestamp3);
-		await Reading.insertAll([reading1, reading2]);
+		await Reading.insertAll([reading1, reading2], conn);
```
I don't know the logic but want to be sure conn is properly set in this code and next.

huss commented on src/server/test/db/loadFromCsvStreamTests.js:
```diff
@@ -90,10 +90,12 @@ mocha.describe('Read mamc log from a file: ', () => {
 				return new Reading(meter.id, readRate, startTimestamp, endTimestamp);
-			}, (readings, tx) => Reading.insertOrIgnoreAll(readings, tx));
+			}, (readings, tx) => Reading.insertOrIgnoreAll(readings, tx),
+			conn
```
Again, I'm no expert on this logic but want to check conn is set here and below.

huss reviewed (approved):
I tried to look this over for obvious concerns. Only a few comments. I'm taking the unusual step of approving so we can move forward with 0.5. However, I still want someone else to carefully go through these changes before we go into production.

NoraCodes committed:
Address PR comments re documentation

NoraCodes committed:
Remove rogue (failing) duplicate tests

huss reviewed (approved):
Seems fine.

NoraCodes committed:
Merge fast-pt-compression.

kuchtact commented on src/server/log.js:
```diff
@@ -48,24 +49,36 @@ class Logger {
+		if (this.logToFile) {
+			fs.appendFile(logFile, messageToLog, err => {
+				if (err) {
+					console.error(`Failed to write to log file: ${err}`); // tslint:disable-line no-console
```
Do we want to also write out the stack trace just like in line 102 of this file?

kuchtact commented on src/server/log.js:
```diff
@@ -141,9 +147,17 @@ class Logger {
 const defaultLogger = new Logger(logFile);
 
+/*
+ * Wherever logging is available, the process should log unhandled rejections.
```
Maybe add a comment on what exactly the 'process' is. I'm not sure when looking at `process.on(...` what exactly `process` is. I'm sure it is something obvious a little more on what exactly that is (whether it be something we are importing or something else) might be useful.

kuchtact commented on src/server/migrations/migrateDatabase.js:
```diff
@@ -169,13 +169,17 @@ async function migrateDatabaseTransaction(neededFiles, allMigrationFiles) {
+async function migrateAll(toVersion, migrationItems, conn) {
+	const currentVersion = await Migration.getCurrentVersion(conn);
+	if (compareSemanticVersion(currentVersion, toVersion) === 0) {
+		throw new Error('You have the highest version');
```
Do we not want people to be able to roll back?

NoraCodes replied:
This is not possible as we don't have a mechanism for down migrations. Eventually we definitely need to change that.

kuchtact commented on src/server/models/Baseline.js:
```diff
@@ -14,15 +13,15 @@ class Baseline {
 	}
-	static createTable() {
-		return getDB().none(sqlFile('baseline/create_baseline_table.sql'));
+	static createTable(conn) {
```
Maybe add block comments to each of these methods. They will probably be a bit redundant but it will keep with the code style of the other `*.js` files.

kuchtact commented on src/server/models/Group.js:
```diff
@@ -176,28 +176,28 @@ class Group {
 	 * Returns a promise to delete a group and purge all trace of it form the memories of its parents and children
 	 * @param groupID The ID of the group to be deleted
-	 * @param conn The connection to be used. Defaults to (*GASP!*) the default database connection
+	 * @param conn the connection to be used.
```
I really think that the `(*GASP!*)` part of the previous comment needs to stay.

kuchtact commented on src/server/models/Migration.js:
```diff
@@ -25,50 +24,53 @@ class Migration {
 	}
 
-	static async insertDefaultMigration() {
+	static async insertDefaultMigration(conn) {
```
Block comment goes here.

kuchtact commented on src/server/models/Migration.js:
```diff
@@ -25,50 +24,53 @@ class Migration {
 
 	/**
 	 * Returns a promise to create the migration table.
+	 * @param conn is the connection to use.
```
Should we keep the copy paste on the `@param conn the connection to be used.` instead of this?

kuchtact commented on src/server/models/Preferences.js:
```diff
@@ -24,28 +23,37 @@ class Preferences {
-	static async update(newPreferences) {
+	/**
+	 * Update the preferences in the database.
+	 * @param conn the database connection to use.
```
I like how line 42 and this line have different comments for `@param conn`. We could also see how many different ways we could describe what `conn` is. Hmmmm...

kuchtact commented on src/server/models/Reading.js:
```diff
@@ -24,87 +23,109 @@ class Reading {
+	 * @returns {number} the number of readings in the entire readings table
+	 */
+	static async count(conn) {
+		const { count } = await conn.one('SELECT COUNT(*) as count FROM readings');
```
Do we want to have sql stuff actually written in the `.js` file? I know it's a real simple query but to keep with how all the other queries are set up, it might be better to have this as a `.sql` file instead.

kuchtact commented on src/server/models/Reading.js:
```diff
@@ -24,87 +23,109 @@ class Reading {
-	static insertOrUpdateAll(readings, conn = getDB) {
-		return conn().tx(t => t.sequence(function seq(i) {
+	static insertOrUpdateAll(readings, conn) {
+		return conn.tx(t => t.sequence(function seq(i) {
```
Can we give `seq` name that better explains what it is? It seems like `seq` just means sequence but I'm not sure why that is the naming for what this does.

kuchtact commented on src/server/models/Reading.js:
```diff
@@ -162,11 +183,11 @@ class Reading {
-	static async getCompressedReadings(meterIDs, fromTimestamp = null, toTimestamp = null, numPoints = 500, conn = getDB) {
-		const allCompressedReadings = await conn().func('compressed_readings',
+	static async getCompressedReadings(meterIDs, fromTimestamp = null, toTimestamp = null, numPoints = 500, conn) {
+		const allCompressedReadings = await conn.func('compressed_readings',
```
Why does this use `func` instead of doing an sql file call. I'm fine with how it is, maybe just a comment to explain why it is different.

kuchtact commented on src/server/routes/baseline.js:
```diff
@@ -2,29 +2,32 @@
-			req.body.note
-		);
-		await baseline.insert();
+			req.body.note);
```
I'm not sure if I like having `);` at the end of a line instead of on a new line. The nice thing about having it on a new line is that it makes it real easy to see all the params entering `Baseline`.

kuchtact commented on src/server/routes/readings.js:
```diff
@@ -52,14 +53,15 @@ router.get('/line/meters/:meter_ids', async (req, res) => {
 		const timeInterval = TimeInterval.fromString(req.query.timeInterval);
 		try {
-			const rawCompressedReadings = await Reading.getCompressedReadings(meterIDs, timeInterval.startTimestamp, timeInterval.endTimestamp, 100);
+			const rawCompressedReadings = await Reading.getCompressedReadings(meterIDs, timeInterval.startTimestamp, timeInterval.endTimestamp, 100, conn);
```
Should we make the function then params into two lines just like on line 110?

kuchtact commented on src/server/routes/readings.js:
```diff
@@ -193,16 +198,18 @@ router.get('/bar/groups/:group_ids', async (req, res) => {
 		}
 	};
-	if (!validate(req.params, validParams).valid || !validate(validQueries, req.query).valid) {
+
```
Do we want this extra line? The other places where a new validation if statement is do not have the extra space.

kuchtact commented on src/server/services/addMamacMeters.js:
```diff
@@ -5,15 +5,28 @@
+			const errors = await insertMetersWrapper(absolutePath, conn);
+		} catch (errors) {
+			for (const err of errors) {
+				log.error(`Error inserting meters: ${err}`, err);
```
Should we also add stack traces? Maybe just once since they are all similar? I'm not sure.

kuchtact commented on src/server/test/common.js:
```diff
@@ -0,0 +1,92 @@
+	// They should be, since they were all created during a previous test.
+	await conn.none('DROP OWNED BY current_user;');
+	await createSchema(conn);
+	await testUser.insert(conn);
```
Add a comment to what `testUser.insert(conn)` does since it is not immediately obvious to me.

kuchtact commented on src/server/test/db/barchartReadingsTests.js:
```diff
@@ -26,18 +16,23 @@ mocha.describe('Barchart Readings', () => {
+		// TODO: This is awkward. Is there a better way to have conn as a non-optional
+		// without requiring changing EVERY call to getBarchartReadings?
+		// If not, well, we need to change every call.
+		const barchartReadings = await Reading.getBarchartReadings([meter.id], moment.duration(2, 'h'), null, null, conn);
```
I think it is just fine to change all of these calls because then it is easy to see where exactly `conn` is used.

NoraCodes replied:
What I meant here was to eliminate the `null, null` near the end.

kuchtact replied:
Sounds good. I can tell you I have absolutely zero clue how to do that.

kuchtact commented on src/server/test/db/compressedReadingsTests.js:
```diff
@@ -103,39 +101,40 @@ mocha.describe('Compressed Readings', () => {
+			await groupB.adoptMeter(meter2.id, conn);
+			await groupC.adoptGroup(groupA.id, conn);
+			await groupC.adoptGroup(groupB.id, conn);
 			const reading1 = new Reading(meter.id, readingValue, startTimestamp, endTimestamp);
```
I think there should be a bit of spacing between adding the meters and then adding the readings. Makes for easier... reading.

kuchtact commented on src/server/test/db/groupTests.js:
```diff
@@ -4,98 +4,96 @@
 		let larry = new Group(undefined, 'Larry');
-		await larry.insert();
+		await larry.insert(conn);
 		// pull larry back out of the db so that we get his ID
```
Capitalization on `p` in `Pull`. Does `larry` also need to be capitalized?

kuchtact commented on src/server/test/db/groupTests.js:
```diff
@@ -104,87 +102,90 @@ mocha.describe('Groups', () => {
+			conn = testDB.getConnection();
+			const parent = await Group.getByName('A', conn);
+			const lovedChild = await Group.getByName('B', conn);
+			const impendingOrphan = await Group.getByName('C', conn);
```
... these variable names ...

kuchtact commented on src/server/test/db/insertMamacReadingsFromCSVTests.js:
```diff
@@ -4,56 +4,52 @@
-		await new Meter(undefined, 'Meter', null, false, Meter.type.MAMAC).insert();
-		meter = await Meter.getByName('Meter');
+		conn = testDB.getConnection();
+		await new Meter(undefined, 'Meter', null, false, Meter.type.MAMAC).insert(conn);
```
When similar code in `compressedReadingsTests.js` was in a transaction. Do we want this as a transaction or not?

kuchtact commented on src/server/test/db/insertMamacReadingsFromCSVTests.js:
```diff
@@ -4,56 +4,52 @@
 			expect.fail('should have thrown an exception');
 		} catch (e) {
 			// We want this to error
+			// TODO: I think Mocha actually has a way to do this with expect()
```
Did we find a way to do this?

kuchtact commented on src/server/test/db/metasysTests.js:
```diff
@@ -0,0 +1,74 @@
+		mocha.beforeEach(async () => {
+			conn = testDB.getConnection();
+			await new Meter(undefined, 'metasys-duplicate', null, false, Meter.type.METASYS).insert(conn);
+		//	meter = await Meter.getByName('metasys-duplicate');
```
Why is this commented out?

kuchtact commented on src/server/test/db/metasysTests.js:
```diff
@@ -0,0 +1,74 @@
+		mocha.it('errors correctly on an invalid file', () => {
+			conn = testDB.getConnection();
+			const testFilePath = path.join(__dirname, 'data', 'metasys-invalid.csv');
+			return expect(readMetasysData(testFilePath, 30, 1, false, conn)).to.eventually.be.rejected;
```
Why does this expect `return` instead of `await`?

NoraCodes replied:
We get better error messages by giving the promise back to Mocha here.

kuchtact commented on src/server/test/db/migrations.js:
```diff
@@ -84,41 +75,46 @@ mocha.describe('Migrations', () => {
-			await clearMigrationsTable();
+			const conn = testDB.getConnection();
+			await clearMigrationsTable(conn);
 			//await new Migration(undefined, '0.0.0', '0.100.0');
```
Why is this commented out?

kuchtact commented on src/server/test/migration/migrateDatabaseInvalidTests.js:
```diff
@@ -0,0 +1,60 @@
+		fromVersion,
+		toVersion,
+		up: async dbt => {
+			// migration here
```
What does `migration here` mean?

kuchtact reviewed (changes requested):
Pretty good overall. A couple possible changes to make but nothing major.

NoraCodes committed:
Merge meters editing

NoraCodes committed:
Merge compare with shift

NoraCodes committed:
Fix TSLint using the global config file always

NoraCodes committed:
Fix migration for 0.3 to 0.4 upgrade.

huss commented:
I think this is where we are posting comments on the big merge for 0.5.

I've only done limited testing on the server but here are some items I have noticed:

- When you use compare you often get "Select one or more items to compare". At other times it does not display all the meters/groups but only does one. It also does not display items if you go from line or bar to compare. I cannot get any group to display in compare.

- If you try to use a chart link, the dashboard pops up for a fraction of a second and then goes away. You are now at the main URL for the site.

- (not a bug and not new but noting) If you try to use a select all in the chart link it selects many items on the dashboard. Given the length, it would be nice to only get the link. Given the layout this might be tricky (I don't know) so not a priority, esp. if difficult.

- If you go to the new groups page and try to make a change it correctly asks for a login. The meters page puts a message on the side about a failure (not told why). In the long-term we should be consistent. If not hard, maybe we should hide the buttons/features you cannot do if not logged in (we should discuss)

- I tried to edit on meters to test. The save acted weird and now I cannot see any meters. I closed/opened the page and it came back/worked. I'm not sure what caused it but something seems off. Finally, I unsure what disable meter does since it did not change the meters displayed.

- Have we tested HTTPS? I know some other changes are needed with the code. At this point it does not work on the server.

NoraCodes commented:
> When you use compare you often get "Select one or more items to compare". At other times it does not display all the meters/groups but only does one. It also does not display items if you go from line or bar to compare. I cannot get any group to display in compare.

I think this is a known bug with compare. If not, I'm not sure what caused it, as I didn't change any front-end code from the compare with shift PR.

> If you try to use a chart link, the dashboard pops up for a fraction of a second and then goes away. You are now at the main URL for the site.

This is definitely a bug, I'll do my best to squash it tonight.

> I tried to edit on meters to test. The save acted weird and now I cannot see any meters. I closed/opened the page and it came back/worked. I'm not sure what caused it but something seems off.

I'll try to reproduce this on my machine, I currently can't but have some theories.

> Finally, I unsure what disable meter does since it did not change the meters displayed.

Disabled meters are not updated when `updateMamacMeters` is run. This terminology is from the pre-meters-editing database but could probably be made more clear.

> Have we tested HTTPS? I know some other changes are needed with the code. At this point it does not work on the server.

As I've mentioned before, we can't enable HTTPS through just code changes; it needs a reworking of how the web server (Apache) presents our sites, and development (3030) currently does not have a subdirectory associated with it.

> If you go to the new groups page and try to make a change it correctly asks for a login. The meters page puts a message on the side about a failure (not told why). In the long-term we should be consistent. If not hard, maybe we should hide the buttons/features you cannot do if not logged in (we should discuss)

This is happening because you have an existing token which has timed out. This is a bug but will not happen for people who have never been logged in; the buttons are hidden for them.

Many thanks!

NoraCodes committed:
Address Cameron's PR comments.

kuchtact commented on src/server/log.js:
```diff
@@ -119,7 +119,7 @@ class Logger {
 	 * @param {String} message the message to log
 	 * @param {Error?} error An optional error object to include information about
-	 * @param {boolean?} skipMail Don't e-mail this message even if we would normally emit an e-mail for this level.
+	 * @param {boolean?} skipMail Don't e-mail this message eveBland error messagen if we would normally emit an e-mail for this level.
```
What happened here. This comment went wacko.

NoraCodes replied:
Oh dear. I must have accidentally hit paste.

NoraCodes committed:
Fix bug where chart links went to a blank page

The whole permalink, including the query string, was being used as the
base URL, because the base tag's href was blank. Now, the server injects
a slash if a relative URL isn't set.

NoraCodes committed:
Remove erroneous nonsense in comment

huss reviewed (approved):
Seems fine.

NoraCodes committed:
Don't send e-mails during every startup

huss reviewed (approved):
Fine. Just for the record, this fixes errors logged during the changeover.

NoraCodes committed:
Fix lint errors

NoraCodes committed:
Fix preferences not updating through API

The model-internal use of Preferences.get() was not correctly using
the connection object.

NoraCodes committed:
Fix users with expired JWTs thinking they had auth

Now, whenever you go to an authenticated page, even if it's not
required, the JWT is checked with /api/verification and you're kicked
back to the login screen if it's not acceptable.

NoraCodes committed:
Fix issue where login redirected to /, not baseURL

NoraCodes committed:
Normalize all browserHistory usages to use baseURL

Now, the login page redirects properly.

NoraCodes committed:
Add missing header
